### PR TITLE
async-41: Better Octree Performance

### PR DESCRIPTION
**Pull request by pwinston**
# Description
* Add canceling in-progress loads if no longer in view.
    * We had this for single-scale but never octree. 
* Add priority scheme to load 3, 2 and 1 level up first.
* Fix "double poll" issue in `QtPoll` with new `IntervalTimer`.
* Add event so visual can "wake up" QtPoll when new data is loaded.
    * Because new data might be loaded "out of the blue"
    * With no camera movement or loading happening at all.
    * But we need to start polling to load the chunks into VRAM.

# Demo 1

`NAPARI_OCTREE=1 napari https://s3.embassy.ebi.ac.uk/idr/zarr/v0.1/4495402.zarr`

Which requires `pip install ome-zarr`.

napari-octree

This dataset takes like 1-2 seconds to load a single tile. So if you move enough, you have to wait that long since all workers will be loading stale data. We can cancel queued loads, but not ones already in progress, at least today.  It's impossible to cancel arbitrary workers, but we might be able to cancel network requests but it won't be easy.

There are still some bugs where tiles don't load, but the  "cancel in-progress loads" and "priority loading" are at least roughly working now.

# Demo 2

`NAPARI_OCTREE=1 napari http://hyper-resolution.org/dzi/Rijksmuseum/SK-C-5/SK-C-5_VIS_20-um_2019-12-21.dzi`

Which requires `pip install napari-dzi-zarr`.

This dataset loads in like 250ms per tile.

napari-octree-3

# Demo 3

napari-octree-2

# Webmon

* Much better frame rate plot in https://github.com/pwinston/webmon/pull/21

frame-rate-2

# Type of change
- [x] New feature in experimental code

# How has this been tested?
- [x] Manual testing

### CODE REVIEW

**Comment by pwinston** [comment]
Finally figured out a source of the slow loading in the astronaut pyramid. We have logic that says "if you are consistently loading chunks in under 30ms, then they are fast enough to load synchronously."

We implemented that long before the octree! It was for single-scale time series. Where we figured why use the `ChunkLoader` if the loads can happen at 30Hz or better. I think we had a case where loads only took 10ms. So loading sync actually made it faster than using threads.

So this in case you see the chunk took 24.532ms to load. So under 30ms. So we switch to synchronous. But then we switched octree levels and loaded 34 chunks in a row, synchronously. Around 30ms each for a total glitch of just over 1 second! No wonder it was hanging.

I think we should just disable this "auto sync" feature for octree. Since we are loading chunks/tiles I can't really see a reason to load them sync. If they already ndarray they will be sync. But this "auto sync" was for a Dask array that just happened to be really fast, maybe it was Dask but in-memory. But we can add that back if need. I think always async is better for octree for now.

Screen Shot 2020-12-12 at 11 49 48 PM

**Comment by pwinston** [comment]
I have that turned off locally and that gets rid of the huge pauses, which is great. But I'm still seen 80+ms frames. On a whim, I cut it down from 6 workers to 1 worker, and then it's completely smooth 60Hz. 

Setting the number of workers automatically is something we have not ooked into at all. It's been on 6 since almost day one, I've never really experimented. So if we really do need 1 sometimes and 6 sometimes, we have to figure that out. We can give the user the option, but we really want a pretty good "automatic" mode so the user doesn't have to mess with it.

It could be something else about having 6 though. Need to experiment.

With one worker, this "long" frame is only 15ms, some are 12ms:

Screen Shot 2020-12-13 at 12 14 04 AM

**Comment by pwinston** [comment]
With 1 worker the loads are slower than the draws, but they are still fast, here a load is 35ms:

Screen Shot 2020-12-13 at 12 19 13 AM

With 4 worker the draws get slower (50+ ms) but the loads get even slower, here over 200ms:

Screen Shot 2020-12-13 at 12 20 29 AM

In this case the zarr file is coming off SSD.  It's possible it's very CPU-intensive and with the GIL we are just fighting with ourselves. I don't want to go back to processes. I think we can dynamically change the number of workers. Basically if one is enough stop with one. Only go up if it's not increasing the load time of a single chunk.

If it's CPU-bound then every worker you add makes it worse. If it's high latency IO more workers should help, more simultaneous downloads. But that needs to be proven out.

**Comment by pwinston** [comment]
@sofroniewn I checked this in with "auto sync" disabled and 1 worker thread. If you can try it out. Together these give me 60Hz with the astronaut zarr file off SSD. I think this will close #1942 if it's working well.

This does not yet have the priority queue of which tiles to load, all the tricks/ordering we've talked about.

**Comment by sofroniewn** [comment]
> @sofroniewn I checked this in with "auto sync" disabled and 1 worker thread. If you can try it out. Together these give me 60Hz with the astronaut zarr file off SSD. I think this will close #1942 if it's working well.

Great catch, I should be able to try the PR this afternoon!

**Comment by pwinston** [comment]
Thinking about it we actually could have a sync mode for octree. But we'd need to limit it to only N chunks per frame.  

So the old way was "if loads are fast, switch to sync mode" and all future are sync, unless it gets slow.

The new way would be "if loads are fast, switch to sync mode", but in sync mode only load N chunks synchronously per frame. Where N is how many chunks you can load within some budget like 20ms.

The nice thing with loading in RAM is you could meter it pretty accurately. Loading into VRAM it's harder to predict the time cost, because much of the cost of loading into VRAM doesn't happen until it calls `glFlush` is called, when the card transfers the texture and does whatever it does.

At any rate this is down the details. But it might pay off for super fast data like off SSD.

**Comment by sofroniewn** [comment]
Dropping "auto sync" is fine, but performance feels much better with `6` workers for a dataset like `napari http://hyper-resolution.org/dzi/Rijksmuseum/SK-C-5/SK-C-5_VIS_20-um_2019-12-21.dzi` which I guess is pure IO bound as it's fetching remote data.

With only one worker sometimes after this change it can feel like it just stops loading new tiles in an intermediate state, until I move the mouse again (maybe it would eventually catch up, but it feels like a very long time). It doesn't currently do that on master, or if I bump the number of workers back up to `6` on this PR.

Screen Shot 2020-12-13 at 4 10 29 PM

Up to you how we move forward

**Comment by pwinston** [comment]
Well it shouldn't hang. So that could be a bug. Although that does suggest someday we might have to deal with killing and restarting loads. But probably it's a bug. 

For remote datasets more workers will usually be better. For S3 transfers we tended to use 10. The exact optimal number is very hard to know. But 1 is not it for remote. 

Ideally we can automatically and dynamically change the number of workers from 0 (sync) to 1 to MAX. And least those three steps if not let it dial in any number from 1 to MAX. 

I'll have to think about it how to make it automatic. We can also put it in the GUI somewhere. At least while we are still learning.

So I guess I'd wait on merging to think about it a little more.

**Comment by pwinston** [comment]
Above is in-progress / debugging, but working on clearing/cancelling any in-progress loads for chunks no longer in view. We had this for pre-octree images, but just never implemented it for octree. It led to chunk loading getting really backed up when viewing remote data, sometimes by 30 seconds.

Fixing this will also help local data. Although it's already very fast, so this should fly.

Also I set the workers up to 10. The real problem that led to setting it to 1 was the "dask level 0" problem discussed at length in #2012. Although basically any load that requires a lot of CPU will be made worse with more workers.

At the very least we need UI for setting the number of workers, probably in a menu. Automatically changing the number of works from 1 to 10 would be ideal. But it's hard to think of a reliable way to do that.  You REALLY don't want to set it wrong or you are making things load much slower than is needed.

**Comment by pwinston** [comment]
Basically, with CPU-intensive loads, the more workers you have the worse things will perform. Whereas with loads mostly blocked on IO, the more workers the better. Up to saturation of your bandwidth or not wanting to spam the remote server too much.

There are ways to get a thread's CPU usage. But the APIs are kind of obscure and I'm not sure they will work on all platforms. But that might work. If the workers are mostly blocked on IO the thread CPU usage should be almost zero. Whereas with heavy CPU it should be really high.

We always used 10 simultaneous downloads for S3, which I think we used in a lot of examples.

But maybe if we can fix #2012 there will be no more CPU-intensive loads for mainstream use. And for ML computations and the like, maybe it's okay to expect the caller to set a flag. Just like they have to turn the cache off. Or we get automatic worker tuning working reliably on all platforms.

**Comment by sofroniewn** [comment]
Just gave this a try. Performance is much much better, really very noticeably so, and the level priority stuff feels very good.

> There are still some bugs where tiles don't load, but the "cancel in-progress loads" and "priority loading" are at least roughly working now.

This did happen a couple times to me too, see example below, but it's not too frequent, and it cleans up if you start moving to new set of tiles.
 
Screen Shot 2020-12-16 at 9 16 09 PM

I also sometimes think I detect slight shifts (not more than a pixel or two) when transitioning between levels that seem to appear transiently and then correct. It might be more noticeable with the grid on, and I'm not sure if it isn't my eyes playing tricks on me, but thought I'd mention it.

I also get segfaults most of the time when I quit with the octree on, but I don't think that is new to this PR. I could go back in time and see when it first started happening if you never see that.

Nice work!

File: napari/_vispy/experimental/vispy_tiled_image_layer.py
Comment (sofroniewn) [inline napari/_vispy/experimental/vispy_tiled_image_layer.py:90]: I don't love that we had to do this, but I see why it might be necessary. There had been talk of using more events to do communication from the vispy visual a while back, but we didn't end up needing it. We might go down that path again though, but can cross those bridge when we get to them!

**Review by sofroniewn — APPROVED** [review]
I'm approving. Changes in performance are very noticeable. I flagged the addition of the event emitter to the vispy visual which I think makes sense, but isn't something we needed before.

This PR is still in draft, so I'll wait to hear more from @pwinston before merging, but it looks good!

File: napari/_vispy/experimental/vispy_tiled_image_layer.py
Comment (pwinston) [reply napari/_vispy/experimental/vispy_tiled_image_layer.py:90]: The list of things that would go away if were just drawing constantly is growing. But I still like the idea that napari can go 100% quiet. But there might be some global way to achieve that. Here we are trying to be super precise so napari won't poll or draw even a single frame more than it has to.

Maybe there's a way to just draw all the time "if the window is active" and "the user is doing something". As long as it also keeps drawing until all loads have finished, even if the user is not doing anything. That's the main source of complexity. Prior to octree we only drew when the camera moved. Now loading is a type of "animation" that needs to finish even if the user has gone to get coffee.

But maybe we could have some global system. And anyone can just ping it say "keep going". Things do seem to be getting complicated just to workaround the issue that we are not always drawing.

**Comment by sofroniewn** [comment]
Ah and we do have an async only test fail that should be fixed before merge too

File: napari/_vispy/experimental/vispy_tiled_image_layer.py
Comment (sofroniewn) [reply napari/_vispy/experimental/vispy_tiled_image_layer.py:90]: > Things to do seem to be getting complicated just to workaround the issue that we are not always drawing.

Yeah, definitely something to consider, but agreed I do like that it can "go 100% quiet" too ....

**Comment by pwinston** [comment]
Yes, I want to go through the diff looking for rough spots tomorrow. And maybe fix some bugs, but if not get it merged then start a new PR to fix.

One neat thing not really documented is you can set 2 log files now:

```
{
    "loader": {
        "force_synchronous": false,
        "num_workers": 20,
        "use_processes": false,
        "auto_sync_ms": 30,
        "delay_queue_ms": 100,
        "log_path": "/Users/pbw/loader.log"
    },
    "octree": {
        "enabled": true,
        "tile_size": 256,
        "preload": {"level": 2, "level+1": 3, "level+2": 4},
        "log_path": "/Users/pbw/octree.log"
    }
}
```

It was useful because they were both scrolling a lot, but at different times. So you kind of wanted both.

**Comment by pwinston** [comment]
I hopefully fixed the stuck tiles. What was happening was if we canceled a load, we couldn't then load that chunk again later. Now we can. 

We don't handle an actual network request that just never completes. We'll have to think about that. We could start a new load on that same chunk. Although we don't today. Whether we do another load or not, the threading loading will potentially be hung forever. This is bad because then we have one less worker thread, forever. You'd hope the load is smart enough to time out. But we are just accessing a Dask array so we don't know.

There might be an argument for using asyncio there. Maybe threads and then each one has a pool of asyncio tasks. Because I think with asyncio tasks if one hangs forever we don't care. It can just it there and we can still make new tasks. But using asyncio would require a lot of new R&D.

**Comment by pwinston** [comment]
## I'm seeing unexplained crashes with PyQt5 so I'm using PySide2. 

I'm not altering the dask chunk size and not changing the scene graph, except to add the visual once. I can't rule out there's not an octree bug between events and polling and things. But certainly, there's a clear pattern now that many different things can cause PyQt5-only crashes. So we can keep banging our heads on the PyQt5 issues, but in order to focus on rendering I'm using PySide2 for now.

**Comment by sofroniewn** [comment]
Makes sense to focus on PySide2. This is now very very nice. I no longer get any stuck tiles. I do still see this transient shift in load, it's not a huge deal, but I do wonder if it's just a couple pixel offset off somewhere and without it things would look flawless. See this gif and look at the tile near the bottom right corner with the man's white trousers and see how it appears to shift ever so slightly during the zoom out, but then corrects. I also see something similar in my pathology dataset so I do think they are doing something weird downsampling.

octree_zoomout

My guess is that it relates to that in napari we consider the center of our top left pixel to be at (0, 0) in a "world coordinate" but vispy considers the top left corner of the top left pixel to be at (0, 0), and so we need to do an offset here https://github.com/napari/napari/blob/v0.4.2/napari/_vispy/vispy_base_layer.py#L131-L141 and that now we're in the mutliscale/ tiled space that offset or other offsets might not quite be happening right, but I'm not sure.

See this image from https://github.com/InsightSoftwareConsortium/SimpleITK-Notebooks/blob/master/Python/03_Image_Details.ipynb for a little more about image origins / pixel vs world coordinates (but note that our y-axis inverted relative to theirs)

image

**Comment by pwinston** [comment]
I added a 10% expansion factor to the view. So that we will sometimes load tiles (from all 3 levels) that are slightly outside the view. Need to make this config option, but it helps.

Also given that, we should add sorting distance from center. So that will automatically leave the ones fully outside the view, the padding, until the end. Will wait for next PR for that. And for the pixel shift I can figure that out.

I will check over the diffs now.

**Comment by pwinston** [comment]
Okay marking this ready.

One thing I will do soon is go through all the `TODO_OCTREE` comments and either remove them or make sure they are explained clearly. Some were just for me, others should remain for future reference.

**Comment by pwinston** [comment]
One thing that might or might not be slowing us down is `OctreeImage.on_chunk_loaded` might be called several times per frame. I think. But it signals `events.loaded()` each time. And when that happens the visual asks the layer for the currently drawable chunks.

 I'd like to batch them up somehow. But it's coming from `QtChunkReceiver` and it's not obvious how to do it. Might involve `QtPoll`! It really makes more sense to have `on_chunks_loaded()` which is given N chunks (at least 1) and then it shoves them in and calls `events.loaded()` just once. Then the visual calls `layer.get_chunks` just once, and it has all the newly loaded chunks.

This pattern of switching from 1 at a time to batch of N comes up all the time.  It's tricky because the slower way works fine. And you may not realize it can be improved by batching.

**Comment by sofroniewn** [comment]
Ok, all that sounds good. Expansion factor looks good too. I will merge now.

**Pull request merged**
